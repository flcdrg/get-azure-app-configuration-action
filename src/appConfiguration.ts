import { PagedAsyncIterableIterator } from '@azure/core-paging';
import * as core from '@actions/core';
import * as io from '@actions/io';
import {
  AppConfigurationClient,
  ConfigurationSetting,
  ListConfigurationSettingPage,
  ListConfigurationSettingsOptions,
  PageSettings
} from '@azure/app-configuration';
import { executeAzCliCommand } from './executeAzCliCommand';

export async function getKeys(
  resourceGroup: string,
  appConfigurationName: string,
  filter: ListConfigurationSettingsOptions
): Promise<
  PagedAsyncIterableIterator<
    ConfigurationSetting<string>,
    ListConfigurationSettingPage,
    PageSettings
  >
> {
  const azPath = await io.which('az', true);
  const connectionString = await executeAzCliCommand(
    azPath,
    `appconfig credential list -g ${resourceGroup} -n ${appConfigurationName} --query "([?name=='Primary Read Only'].connectionString)[0]"`
  );

  core.setSecret(connectionString);

  const client = new AppConfigurationClient(connectionString);

  return client.listConfigurationSettings(filter);
}
