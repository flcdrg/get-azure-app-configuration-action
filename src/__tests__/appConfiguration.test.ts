/**
 * Integration Tests
 *
 * These tests require access to an Azure Subscription which has an App Configuration RESOURCE
 * within a RESOURCE_GROUP. The values for these default to those used in the Azure Subscription
 * used when testing in the GitHub Action(s).
 *
 * The contents of the App Configuration resource should mirror the values stored in the
 * appconfig-sample*.json files (included in this directory)
 *
 * @group integration
 */
import {
  ConfigurationSetting,
  ListConfigurationSettingPage,
  PageSettings
} from '@azure/app-configuration';
import { PagedAsyncIterableIterator } from '@azure/core-paging';
import { getKeys } from '../appConfiguration';

type StringDictionary = {
  [id: string]: string;
};

describe('appConfiguration', () => {
  const resourceGroup =
    process.env.RESOURCE_GROUP ?? 'rg-appconfig-test-australiaeast';
  const resource = process.env.RESOURCE ?? 'appcs-appconfig-test-australiaeast';

  it('all', async () => {
    const keys = await getKeys(resourceGroup, resource, {
      keyFilter: '',
      labelFilter: ''
    });

    const result: StringDictionary = {};

    await copyToResult(keys, result);

    expect(result).toMatchSnapshot();
  });

  it('no labels', async () => {
    const keys = await getKeys(resourceGroup, resource, {
      keyFilter: '*',
      labelFilter: '\0'
    });

    const result: StringDictionary = {};

    await copyToResult(keys, result);

    expect(result).toMatchSnapshot();
  });

  it('nulls', async () => {
    const keys = await getKeys(resourceGroup, resource, {
      keyFilter: undefined,
      labelFilter: undefined
    });

    const result: StringDictionary = {};

    await copyToResult(keys, result);

    expect(result).toMatchSnapshot();
  });

  it('single key1', async () => {
    const keys = await getKeys(resourceGroup, resource, {
      keyFilter: 'key1'
    });

    const result: StringDictionary = {};

    await copyToResult(keys, result);

    expect(result).toMatchSnapshot();
  });

  it('wildcard key*', async () => {
    const keys = await getKeys(resourceGroup, resource, {
      keyFilter: 'key*'
    });

    const result: StringDictionary = {};

    await copyToResult(keys, result);

    expect(result).toMatchSnapshot();
  });

  it('multiple key1,key2', async () => {
    const keys = await getKeys(resourceGroup, resource, {
      keyFilter: 'key1,key2'
    });

    const result: StringDictionary = {};

    await copyToResult(keys, result);

    expect(result).toMatchSnapshot();
  });

  it('single label2', async () => {
    const keys = await getKeys(resourceGroup, resource, {
      labelFilter: 'label2'
    });

    const result: StringDictionary = {};

    await copyToResult(keys, result);

    expect(result).toMatchSnapshot();
  });

  it('wildcard label*', async () => {
    const keys = await getKeys(resourceGroup, resource, {
      labelFilter: 'label*'
    });

    const result: StringDictionary = {};

    await copyToResult(keys, result);

    expect(result).toMatchSnapshot();
  });

  it('key3 and label3', async () => {
    const keys = await getKeys(resourceGroup, resource, {
      keyFilter: 'key3',
      labelFilter: 'label3'
    });

    const result: StringDictionary = {};

    await copyToResult(keys, result);

    expect(result).toMatchSnapshot();
  });
});

async function copyToResult(
  keys: PagedAsyncIterableIterator<
    ConfigurationSetting<string>,
    ListConfigurationSettingPage,
    PageSettings
  >,
  result: StringDictionary
): Promise<void> {
  for await (const setting of keys) {
    if (setting.value) {
      result[setting.key] = setting.value;
    }
  }
}
