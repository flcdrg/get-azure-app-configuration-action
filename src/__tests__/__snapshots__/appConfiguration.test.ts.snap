// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appConfiguration all 1`] = `
{
  "$dollarkey": "dollarkey value",
  ".leadingdot": ".leadingdot value",
  "coloned:key": "coloned:key value",
  "dotted.key": "dotted.key value",
  "key1": "key1 value",
  "key2": "key2 value",
  "key3": "key3 label3 value",
  "ref kv secret 1": "{"uri":"https://kv-appconfig-test-aue.vault.azure.net/secrets/kv-secret-1"}",
  "slash/separated/key": "slash/separated/key value",
  "underscored_key": "underscored_key value",
}
`;

exports[`appConfiguration key3 and label3 1`] = `
{
  "key3": "key3 label3 value",
}
`;

exports[`appConfiguration multiple key1,key2 1`] = `
{
  "key1": "key1 value",
  "key2": "key2 value",
}
`;

exports[`appConfiguration no labels 1`] = `
{
  "$dollarkey": "dollarkey value",
  ".leadingdot": ".leadingdot value",
  "coloned:key": "coloned:key value",
  "dotted.key": "dotted.key value",
  "key1": "key1 value",
  "key2": "key2 value",
  "ref kv secret 1": "{"uri":"https://kv-appconfig-test-aue.vault.azure.net/secrets/kv-secret-1"}",
  "slash/separated/key": "slash/separated/key value",
  "underscored_key": "underscored_key value",
}
`;

exports[`appConfiguration nulls 1`] = `
{
  "$dollarkey": "dollarkey value",
  ".leadingdot": ".leadingdot value",
  "coloned:key": "coloned:key value",
  "dotted.key": "dotted.key value",
  "key1": "key1 value",
  "key2": "key2 value",
  "key3": "key3 label3 value",
  "ref kv secret 1": "{"uri":"https://kv-appconfig-test-aue.vault.azure.net/secrets/kv-secret-1"}",
  "slash/separated/key": "slash/separated/key value",
  "underscored_key": "underscored_key value",
}
`;

exports[`appConfiguration single key1 1`] = `
{
  "key1": "key1 value",
}
`;

exports[`appConfiguration single label2 1`] = `
{
  "key3": "key3 label2 value",
}
`;

exports[`appConfiguration wildcard key* 1`] = `
{
  "key1": "key1 value",
  "key2": "key2 value",
  "key3": "key3 label3 value",
}
`;

exports[`appConfiguration wildcard label* 1`] = `
{
  "key3": "key3 label3 value",
}
`;
